Sqlite Attach Database

elmpp opened the pull request:
### Description of change

Adds support for SQLite [attach databases](https://sqlite.org/lang_attach.html). This allows database files additional to the default "main" file that TypeORM currently supports. 

This implementation is quite similar to current support for multi-database (postgres, mysql) but without the concept of schema paths. The databases are configured in the same way also - as entity decorators; albeit using filepath strings

The bulk of the complexity comes from platform filepath differences - i felt a big requirement would be that migration files, synchronised file locations etc would be portable between windows/posix. As a result there are a few path utils (with tests) to achieve that. 

Notes:
 - ~Foreign keys aren't supported cross-database so are turned off if any dbs are attached. It could be possible to still support them within the same db. What do you think?~ Foreign keys between tables of same db are included
 - ~i wasn't able to build 'better-sqlite' on my machine so hoping CI will test that~
 - adds `ATTACH` statements to migrations

### Pull-Request Checklist

<!--
  Please make sure to review and check all of the following.

  If an item is not applicable, you can add "N/A" to the end.
-->

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

<!--
  🎉 Thank you for contributing and making TypeORM even better!
-->

nebkat commented on src/metadata/EntityMetadata.ts:
```diff
@@ -861,6 +862,11 @@ export class EntityMetadata {
         }
 
+        if (this.connection.driver instanceof SqliteDriver) {
+            // @todo - perhaps use this driver callback in all cases?
```
At this point lets just abstract `buildTablePath()` for all drivers and pass `tableName`, `schema` and `database`. The function could be made optional also and default to using `tableName` by default.

elmpp replied:
I didn't want to do this outright in my first PR because i'm still unclear the differences between `tableName` and `tablePath`.

I think it needs a bit more thought (and probably needs more consistent escaping strategies in certain places)

nebkat replied:
The only difference should be that `tablePath` contains the full database/schema name, while `tableName` does not. I think drivers have also so far been responsible for escaping, but is there something else I'm missing here?

elmpp replied:
the concept of `tablePath` vs `tableName` seems a little blurry throughout - it differs from here in the `EntityMetadata`, through the `Table` and `TableIndex` instances. Also, the same driver instance can expect differing values based on the platform (the `AbstractSqliteDriver/QueryRunner`). e.g. `sql.js` does not support tablePaths on indexes etc

elmpp replied:
tl;dr; i'd rather not include cross-driver changes in this PR

nebkat commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -85,6 +85,7 @@ export abstract class QueryBuilder<Entity> {
      * QueryBuilder can be initialized from given Connection and QueryRunner objects or from given other QueryBuilder.
      */
     constructor(connectionOrQueryBuilder: Connection|QueryBuilder<any>, queryRunner?: QueryRunner) {
+
```
```suggestion
```

nebkat commented on src/driver/sqlite/SqliteDriver.ts:
```diff
@@ -81,6 +87,29 @@ export class SqliteDriver extends AbstractSqliteDriver {
+     */
+    buildTableName(tableName: string, _schema?: string, database?: string): string {
+
+        if (!database) return tableName
```
Missing semicolons throughout this function (and a few others).

nebkat commented on src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:
```diff
@@ -1149,8 +1167,12 @@ export abstract class AbstractSqliteQueryRunner extends BaseQueryRunner implemen
         });
 
+        const [databaseNew, tableNameNew] = this.splitTablePath(newTable.name)
+        // @ts-ignore // unused var
```
Can `_` be used for unused vars?

nebkat replied:
I checked now the syntax is actually to not put in any variable:

`[ , tableNameNew]`

nebkat commented on src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:
```diff
@@ -1197,4 +1218,18 @@ export abstract class AbstractSqliteQueryRunner extends BaseQueryRunner implemen
+     * tablePath e.g. "myDB.myTable", "myTable"
+     */
+    protected splitTablePath(tablePath: string): [string | undefined, string] {
+        return ((tablePath.indexOf('.') !== -1) ? tablePath.split('.') : [undefined, tablePath]) as [string | undefined, string]
```
Should be careful here, `split('.')` could return more than 2. Better to do something like
```suggestion
        let dotIndex = tablePath.indexOf(".");
        return (dotIndex >= 0) ? [tablePath.substr(0, dotIndex), tablePath.substr(dotIndex + 1)] : [undefined, tablePath];
```

nebkat replied:
Also make sure all strings are in double quotes rather than single, should be picked up by ESLint in IDE?

elmpp commented on src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:
```diff
@@ -1150,7 +1172,12 @@ export abstract class AbstractSqliteQueryRunner extends BaseQueryRunner implemen
-        newTable.name = "temporary_" + newTable.name;
+        newTable.name = this.buildTemporaryTablePath(newTable.name)
+
+        // @ts-ignore // ignore var
```
would be nice if this was allowable?

imnotjames commented on src/connection/BaseConnectionOptions.ts:
```diff
@@ -110,6 +110,11 @@ export interface BaseConnectionOptions {
+    /**
+     * The directory where ormconfig has been read from
+     */
+    baseDirectory?: string;
```
What happens when there's no ormconfig?

Why does this feature operate differently from how the standard database path resolution works for sqlite?

imnotjames commented on src/connection/ConnectionOptionsReader.ts:
```diff
@@ -172,11 +176,10 @@ export class ConnectionOptionsReader {
                 if (typeof options.database === "string" &&
-                    options.database.substr(0, 1) !== "/" &&  // unix absolute
-                    options.database.substr(1, 2) !== ":\\" && // windows absolute
+                    !isAbsolute(options.database) &&
```
The attaching should probable be defined within the connection options for sqlite instead of in random entities so they are in the same place as the main file's database.  The entities could then reference the alias instead of the filename.

imnotjames commented on src/driver/sqlite-abstract/AbstractSqliteDriver.ts:
```diff
@@ -39,6 +44,11 @@ export abstract class AbstractSqliteDriver implements Driver {
+    /**
+     * Any attached databases (excepting default 'main')
+     */
+    attachedDatabases: DatabasesMap = {};
```
can this be protected?

imnotjames commented on src/driver/sqlite/SqliteDriver.ts:
```diff
@@ -139,8 +170,27 @@ export class SqliteDriver extends AbstractSqliteDriver {
-    protected async createDatabaseDirectory(fullPath: string): Promise<void> {
-        await mkdirp(path.dirname(fullPath));
+    protected async createDatabaseDirectory(dbPath: string): Promise<void> {
+        await mkdirp(dbPath);
```
Probably can drop this function and instead use `mkdirp` directly

imnotjames reviewed (changes requested):
We aren't using yarn for this project, please remove references to it.

pleerock commented:
I'll close it since there is no answer from author. Feel free to re-open.

elmpp commented:
@pleerock  Hi there - I must have overlooked the second review of @imnotjames . If I find some time this week to respond is there a decent chance it could be considered? (there been a bit of a wait)

Not sure how to reopen this PR tbh either ?

imnotjames commented:
You should be able to reopen it yourself.

elmpp commented:
I see no option to reopen - only the ability to "reference in a new issue". I'd like to keep the comments etc if possible. Weirdly, i'm also unable to upload images in this reply

elmpp commented:
I have rebased the branch to latest master and added more tests. 

I thought adding commits would give me the option to reopen but looks like not 

I'm going to open another PR with a reference to here. I assume you are filtering closed PR comments 👍
